File: demo/src/main/java/com/juankysoriano/rainbow/demo/sketch/rainbow/forces/ParticleWorld.java
package com.juankysoriano.rainbow.demo.sketch.rainbow.forces;

import com.juankysoriano.rainbow.core.drawing.RainbowDrawer;
import com.juankysoriano.rainbow.core.matrix.RVector;

class ParticleWorld {

    private static final int NUMBER_PARTICLES = 100000;
    private static final int NUCLEUS_DIAMETER = 40;
    private static final float HALF = 0.5f;

    private final Particle[] particles;
    private final Nucleus nucleus;
    private final float[] points;

    static ParticleWorld newInstance(int width, int height) {
        RVector nucleusCoordinates = new RVector(width * HALF, height * HALF);
        Nucleus nucleus = new Nucleus(nucleusCoordinates, NUCLEUS_DIAMETER);
        Particle[] particles = generateParticles(nucleus);

        return new ParticleWorld(nucleus, particles);
    }

    private ParticleWorld(Nucleus nucleus, Particle[] particles) {
        this.nucleus = nucleus;
        this.particles = particles;
        this.points = new float[NUMBER_PARTICLES * 2];
    }

    private static Particle[] generateParticles(Nucleus nucleus) {
        Particle[] particles = new Particle[ParticleWorld.NUMBER_PARTICLES];
        for (int i = 0; i < ParticleWorld.NUMBER_PARTICLES; i++) {
            Particle particle = new Particle();
            particle.resetTo(nucleus);
            particles[i] = particle;
        }
        return particles;
    }

    void update() {
        for (int i = 0; i < NUMBER_PARTICLES; i++) {
            setPointFor(i);
            particles[i].updateWith(nucleus);
        }
    }

    void display(RainbowDrawer rainbowDrawer) {
        rainbowDrawer.background(0);
        rainbowDrawer.strokeWeight(0);
        rainbowDrawer.stroke(255, 70);
        rainbowDrawer.point(points);
    }

    private void setPointFor(int particleIndex) {
        RVector particleLocation = particles[particleIndex].getLocation();
        points[particleIndex * 2] = particleLocation.x;
        points[particleIndex * 2 + 1] = particleLocation.y;
    }

    void moveNucleusTo(float x, float y) {
        this.nucleus.getPosition().set(x, y, 0);
    }
}

File: demo/src/main/java/com/juankysoriano/rainbow/demo/sketch/rainbow/forces/ParticleWorld.kt
package com.juankysoriano.rainbow.demo.sketch.rainbow.forces

import com.juankysoriano.rainbow.core.Rainbow
import com.juankysoriano.rainbow.core.math.RVector

private const val NUMBER_PARTICLES = 200000
private const val NUCLEUS_DIAMETER = 40
private const val HALF = 0.5f

internal class ParticleWorld private constructor(private val nucleus: Nucleus,
                                                 private val particles: Array<Particle>,
                                                 private val points: FloatArray = FloatArray(2 * NUMBER_PARTICLES)) {

    fun updateAndDisplay(rainbow: Rainbow) {
        for (i in 0 until NUMBER_PARTICLES) {
            val (x, y) = particles[i].location
            points[i * 2] = x
            points[i * 2 + 1] = y
            particles[i].updateWith(nucleus)
        }

        rainbow.rainbowDrawer.apply {
            strokeWeight(0f)
            stroke(255, 70f)
            point(*points)
        }
    }

    fun moveNucleusTo(x: Float, y: Float) = this.nucleus.origin.set(x, y)

    companion object {
        fun create(width: Int, height: Int): ParticleWorld {
            val nucleusCoordinates = RVector(width * HALF, height * HALF)
            val nucleus = Nucleus(nucleusCoordinates, NUCLEUS_DIAMETER.toFloat())
            val particles = generateParticles(nucleus)

            return ParticleWorld(nucleus, particles)
        }

        private fun generateParticles(nucleus: Nucleus): Array<Particle> {
            val particles = ArrayList<Particle>(NUMBER_PARTICLES)
            for (i in 0 until NUMBER_PARTICLES) {
                val particle = Particle()
                particle.resetTo(nucleus)
                particles.add(particle)
            }
            return particles.toTypedArray()
        }
    }
}
